### ISSUE TITLE
Gutenberg compatibility for PHP/AJAX Validation

### ISSUE THREAD

**Issue by elliotcondon** [AdvancedCustomFields/acf#113]
## Description
A highly used feature in ACF is our PHP validation powered via AJAX. This allows developers to highly customize the validation process via PHP whist providing convenient feedback to the user when a field value does not meet its requirements.

To do this, we use JS to intercept the "submit" action on the form and stop it. We then send all `$_POST` data via an AJAX request and use the returned JSON to either display errors or re-submit the form.

It is important for UX that we validate the form data only when the user clicks the "publish" button.

## Issues
- [ ] The form "submit" event is never fired, so our AJAX validation is never triggered.

## Questions
- [ ] What can we "hook" into to prevent the post from being submit in order to validate via AJAX and then re-submit?
- [ ] How can we get all the $_POST data about the post (title, content, etc)?
- [ ] How can we re-submit the form after validation is complete?
- [ ] Is there an API for displaying an error message in Gutenberg?

**Comment by noisysocks** [comment]
> Can we work on a solution for validating the form data via AJAX that also keeps the above in mind?

Got it. I don't _think_ this is currently possible but it's a new hook that we can discuss adding over in the Gutenberg project.

To clarify: do we need to hook into when the post is being saved, or when the post is being published?

**Comment by noisysocks** [comment]
My current thinking is that ACF could use [`PluginPrePublishPanel`](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-edit-post/#pluginprepublishpanel) to render a component in the pre-publish panel that:

- Dispatches [`lockPostSaving()`](https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/data-core-editor/#lockpostsaving) when the component is mounted so that the user can not click _Publish_.
- Sends off the AJAX request which validates the current post.
  - If the post is valid, dispatches [`unlockPostSaving()`](https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/data-core-editor/#unlockpostsaving).
  - If the post is invalid, displays some UI which indicates to the user which fields need attention.

Very quick mockup of what this could look like:

![localhost_8888_wp-admin_post-new php desktop 4](https://user-images.githubusercontent.com/612155/50947829-31b48280-14f4-11e9-9f18-05f8250b8e4b.png)

Does that make sense? Would this be a workable approach?

**Comment by elliotcondon** [comment]
Hi @noisysocks 

This is a great idea, but will this also work for updating a published post?

**Comment by noisysocks** [comment]
Hey @elliotcondon. That's a good point. No, unfortunately this won't handle the case where a published post is updated. We may have to add a new filter for this after all. I've suggested this over in https://github.com/WordPress/gutenberg/issues/13413.

**Comment by elliotcondon** [comment]
Hi @noisysocks 

Just following up on this one. Did we get anywhere with a solution to "validate" and "prevent" the form from being submit?

If not yet explored, would it be possible to use a promise object + a JS filter to allow 3rd party (or core) logic to stop/trigger the saving process?

**Comment by noisysocks** [comment]
Hey @elliotcondon! I like your new avatar 🙂

I haven't been actively working on this. https://github.com/WordPress/gutenberg/issues/13413 tracks adding the API which I prefer which is a promise object + JS filter. https://github.com/WordPress/gutenberg/pull/13718 is an unrelated PR that also had a need for such an API. 

@talldan: What's the status on https://github.com/WordPress/gutenberg/pull/13718? Do you think we can include the mentioned API?

**Comment by matgargano** [comment]
Any update here?

**Comment by elliotcondon** [comment]
@matgargano - Not from my end. I'm still waiting for WordPress to add in some sort of action/filter to hook in and customize the saving process.

**Comment by benbowler** [comment]
Is there a ticket on the Wordpress side we can comment on/contribute to?

**Comment by elliotcondon** [comment]
@benbowler - There are a few GitHub threads listed in previous replies on this thread. 

If you're reading this via email, click the GitHub link to view this issue in your browser so you can scroll up and check those links.

For example: https://github.com/WordPress/gutenberg/issues/13413

**Comment by firda-yana** [comment]
Hi all, Do we have any update here?

**Comment by jenilk** [comment]
Hi @elliotcondon,

Any update on this? As the WordPress 5.3 has been released, You probably able to fix the issue now.

Best
Jenil

**Comment by elliotcondon** [comment]
@jenilk I'm not aware of any changes in WP 5.3 that introduced an API for PHP validation. If you can provide some reference, that would be great :)

**Comment by jenilk** [comment]
Hi @elliotcondon  I am also in a same phase as you're. However, I have created a ticket to WordPress where we can all contribute there. As this issue goes to a blocker we need to find the solutions as earliest. 

Ticket: https://core.trac.wordpress.org/ticket/48959

Best 
Jenil

**Comment by adamsilverstein** [comment]
Hi @elliotcondon - although there is no specific API for PHP side validation, _you should be able to achieve something close to what you want with the existing extensibility provided in Gutenberg_ (unless I am missing something):

1. Use [post saving lock](https://github.com/WordPress/gutenberg/pull/10649) to prevent updates/publishing when validation fails, this works for updating or publishing.
2. subscribe to data changes ([example code](https://developer.wordpress.org/block-editor/data/data-core-editor/#lockPostSaving)), or field changes explicitly, then when a field is changed/blurred, send an ajax request to perform PHP validation
3. when everything validates, unlock saving, otherwise keep it locked and display the validation issues. use can use an alert, a sidebar, display the errors inline or some combination of those.

If this doesn't work for your use case, can you explain why? Can you identify exactly what type of hook or filter in GB would enable your use case?

This plugin might be useful to look at: https://github.com/humanmade/publication-checklist

**Comment by elliotcondon** [comment]
Hi @adamsilverstein 

Thanks for the information. The "Post saving lock" API and the example plugin look very promising and I'll begin some testing with this shortly!

**Comment by isuke01** [comment]
Hey, I was using non Gutenberg fo so long, but I decided to use ... some part of if.
Today I tested it, I make it a bit without content, just title ...
![image](https://user-images.githubusercontent.com/12289763/76133734-e45d7380-6019-11ea-9b01-02fa5fe00801.png)
But I noticed this issue ... and I was a bit not happy of it.
So I fixed it.
I tested it on ACF PRO 5.8.8 and WP 5.3.2, I did not test every use case yet. But seems like working, hope it will help someone.
`function enhancement_gutenberg_isu(){
    ?>
    <script type="text/javascript">
        jQuery(document).ready(function($) {
            
            wp.domReady(function(){
                //acf Custom Validation for Gutenberg
                var postSaveButtonClasses = '.editor-post-publish-button';
                $(document).on('click', postSaveButtonClasses , function(e){
                        e.stopPropagation();
                        e.preventDefault();
                        var $button = $(this)
                        acf_validate_fields($button);
                })
            })

            function acf_validate_fields($button){
                var editorInfo = wp.data.select( 'core/editor' );
                var isPublishSidebarOpened = wp.data.select('core/edit-post').isPublishSidebarOpened();
                var tmpButtonText = $button.text();

                wp.data.dispatch( 'core/editor' ).lockPostSaving( 'acfValidateFields' );
                $button.text('Fields validation ...');

                return acf.validateForm({
                    form: jQuery('#editor'),
                    reset: true,
                    failure: function( $form, validator ){
                        var notice = validator.get('notice').data.text
                        wp.data.dispatch( 'core/notices' ).createErrorNotice( notice, { id: 'ACF_VALIDATION', isDismissible: true}  );
                        if(isPublishSidebarOpened){
                            wp.data.dispatch('core/edit-post').closePublishSidebar()
                        }
                    },
                    complete: function( $form, validator ){
                        $button.text(tmpButtonText)
                        wp.data.dispatch( 'core/editor' ).unlockPostSaving( 'acfValidateFields' );
                    },
                    success: function(){
                        //remove notices if there are any from acf validation
                        wp.data.dispatch( 'core/notices' ).removeNotice('ACF_VALIDATION');
                        //save post if is ok to save
                        if(editorInfo.isEditedPostSaveable()){
                            if(!editorInfo.isCurrentPostPublished() && editorInfo.getEditedPostAttribute( 'status' ) === 'draft'){
                                //for some reason if post is draft we must update manually  post status
                                //otherwise post will be saved but not published;
                                wp.data.dispatch('core/editor').editPost({status: 'publish'})
                            }
                            wp.data.dispatch( 'core/editor' ).savePost()
                        }
                    }
                })
            }
        });
    </script>
    <?php
}
add_action('admin_footer', 'enhancement_gutenberg_isu');`
I have learned I can subscribe for editor ... but I think this way is much easier :D

**Comment by elliotcondon** [comment]
Hi @isuke01 

Thank you very much for your contribution here. I too have thought about using a click event listener, but ultimately scrapped the idea as it would not comply with Gutenberg's keyboard accessibility features.

If you are able to find a more holistic way to listen for (and prevent) the "save process", this would be greatly appreciated!

**Comment by isuke01** [comment]
Thans @elliotcondon,

yes, I totally forgot about that thing! I'll look into it.
I'm sure I can solve it using subscribe hook 
`wp.data.subscribe(function () {...})`

I'll post solution if I'm able to fix  :D

**Comment by elliotcondon** [comment]
Hi @isuke01 

Thanks for looking into the keyboard shortcut issue. I'll be sure to test out something similar.
Covering the "click" and "key" events should work for the most part, but I wonder if there are any other situations where the "save" function is called by WP core... perhaps when previewing, etc?

**Comment by isuke01** [comment]
Hi @elliotcondon

Actually one more save of post that should we take care is switch from any status to `private` but not all the way around. For some reason save is called if you switch from "Public" to "Private" only. It also display document alert, I have no clue why it is solved like that, looks strange.

Another thing. "Preview" it does not call save for current post, so no problem but I noticed Preview in this case is not saving any meta. (Tested on Posts, not custom CTP), maybe because fields are not `register_meta`, I don't know. This require more tests.

And about clicking "mod+s" shortcut to save, it works only when something is edited. But actually this does not work on ACF fields, so if you edit something inside ACF only, not in Gutenberg container, you're unable to save post using key commands. BUT! Since in my above function I'm not checking for `isEditedPostDirty()`, it solves this issue 😄

Would be nice to have `pre_save_hook` for Gutenberg but yet, I cant find solution for that. 
Maybe someday! 😄

**Comment by elliotcondon** [comment]
Hi @isuke01 

Thank you for all the extra info!

It looks like we are not the only ones looking for a solution to this problem. Here are some GithUb issues you might be interested in looking over / following:
- https://github.com/WordPress/gutenberg/issues/15568
- https://github.com/WordPress/gutenberg/issues/13413

I agree that the best case scenario would be a `pre_save_hook` for Gutenberg. I wonder if this is something that you might be able to help with? Your knowledge of the Gutenberg editor and API seem very strong.

If you have some free time, please read over those GitHub issues mentioned above, and let me know if you see anything that might be useful 😄.

**Comment by isuke01** [comment]
Hey @elliotcondon 

Thanks, actually my first experience with Gutenberg was 4 days ago, and I just spend few hours on it ... and it was painfully (I hope they will fix docs at some point it is agrrggrr)😄 I just cant deliver to client something that does not work so I tried quick fix it.

Anyway I have find out Gutenberg is not saving meta fields during preview/autosave/auto-draft save. [WordPress/gutenberg/#20755](https://github.com/WordPress/gutenberg/issues/20755).

So basically until I find a way how to solve saving those meta for preview, Gutenberg is not usable for me :/ Thing is, without Gutenberg everything works fine.

Im out of time to solve this issue by now, I'll look into that issues as soon as I can find some time, but for next couple of days Im totally occupied, and simply I'll use classic editor instead for now :)

**Comment by ituki** [comment]
hi @isuke01 

I was so in trouble, tried your code.
It may be a known problem for you,
Paragraph Does not work new line by below code.

```
                //Handle keyshorts
                document.addEventListener('keydown', function(e){
                    //https://developer.wordpress.org/block-editor/packages/packages-keycodes/
                    // remove other bubbling element in DOM
                    e.stopImmediatePropagation()
                    if(wp.keycodes.isKeyboardEvent.primary(e, 's')){
                        saveValidations(e)
                    }
                }, true);
```

**Comment by isuke01** [comment]
@ituki Hey, 
sorry I'm really busy now and I can't do more tests etc, even if I wish :/ 
(my normal work is remote so quarantine does not help me with time, I have even more xD) 

For sure there is more issues, like you should check if event `wp. domReady` exist because it can lead you to problems on non gutenberg pages like options ... you may also check if this is post with gutenberg scree ... to avid code on any other pages than one with gutenberg ...

But to your problem.
Yeh, I see I canceled all input using `e.stopImmediatePropagation()` before check, you may try move it inside check of save shortcode, like 
```
if(wp.keycodes.isKeyboardEvent.primary(e, 's')){
  e.stopImmediatePropagation()
  saveValidations(e)
}
```
So it should cancel only save action.  :)

**Comment by elliotcondon** [comment]
Hi all 👋,

Thank you so much for your support via this thread. There are some great code snippets, ideas and solutions here - they are definitely helping!

I might have an early prototype of a validation solution to share. Its a little rough around the edges, but the main functionality appears to be working well 👍. Feel free to test this out, expand on it and discuss ideas!

```js
var gutenbergValidation = new acf.Model({
	wait: 'ready',
	initialize: function(){
		
		// Customize Gutenberg editor only.
		if( acf.isGutenberg() ) {
			this.customizeEditor();
		}
	},
	customizeEditor: function(){
		
		// Extract vars.
		var editor = wp.data.dispatch( 'core/editor' );
		var notices = wp.data.dispatch( 'core/notices' );
		
		// Reference original method.
		var savePost = editor.savePost;
		
		// Override core method.
		editor.savePost = function(){
			
			// Validate the editor form.
			var valid = acf.validateForm({
				form: $('#editor'),
				reset: true,
				complete: function( $form, validator ){
					
					// Always unlock the form after AJAX.
					editor.unlockPostSaving( 'acf' );
				},
				failure: function( $form, validator ){
					
					// Get validation error and append to Gutenberg notices.
					var notice = validator.get('notice');
					notices.createErrorNotice( notice.get('text'), { 
						id: 'acf-validation', 
						isDismissible: true
					});
					notice.remove();
				},
				success: function(){
					
					// Save post on success.
					savePost();
				}
			});
			
			// Lock the form if waiting for AJAX response.
			if( !valid ) {
				editor.lockPostSaving( 'acf' );
				return false;
			}
			
			// Save post as normal.
			savePost();
		}
	}
});
```

**Comment by Emiliano-Bucci** [comment]
@elliotcondon Hi! It seems to work well! Is there any ETA of including in the plugin core? Thanks! :)

**Comment by elliotcondon** [comment]
@Emiliano-Bucci That's great news! Thanks for testing out this prototype. My plans are to include this in version 5.9.0, which will provide us with a beta + RC workflow allowing developers to test and identify issues with this approach.

5.9.0 is probably a few weeks away from a beta announcement. For now, if this solution works for you, please include it in your theme files until then :)

**Comment by isuke01** [comment]
Looks great! Thanks!

**Comment by adamsilverstein** [comment]
Interesting approach! I didn't realize you could overwrite the editor.save method in that way. Nice work figuring that out :) I'm still a little surprised you weren't able to find a workable solution using the subscribe approach, still I'm happy you have figured out a way to address your use case.

**Comment by benfrem** [comment]
I got a new branch for the testers created with Elliot's code, and it was tested and confirmed as working across about 15 different page templates and two post types. I'd presume our tester tested it with the custom ACF validation hooks we use, too. I'd guess that covered about 30 field groups and about 100 individual fields, including group, image & repeater fields.

I've been emailing with Elliot behind the scenes after we got a similar fix to the above working and tested and I hoped to get it into ACF core (Elliot's similar solution here is probably better), and thought I'd share a related bit. I built similar code for ACF block validation (shown one of the screenshots) that works by doing similar validation during the block's save function and slightly modifying the validate_values PHP function in validation.php to be like the below so it can account for the block prefixes that get POSTed. I guess the good news is it looks like block validation may be implemented somewhat soon / eventually, too. I'm planning to work on reworking the code to just validate on update-button-click, and am excited about how the new validation code is making migrating ACF sites into Gutenberg + ACF/ACF-block sites so much more easy and possible. Thanks!

![Annotation 2020-04-16 133956](https://user-images.githubusercontent.com/19555333/81112901-5c96c700-8eed-11ea-8abd-f004ae50fdec.png)
![Annotation 2020-04-16 133421](https://user-images.githubusercontent.com/19555333/81112902-5c96c700-8eed-11ea-8160-34abeafcc049.png)

function acf_validate_values( $values, $input_prefix = '' ) {
	//print_r($values);
	// bail early if empty
	if( empty($values) ) return;
	//echo "stillgoing";
	///Needs to send only each acf-prefixed post key into the below
	
	if ( 'savepost' == $input_prefix ) {
		foreach ( $values as $postkey => $postvalue ) {
			// echo "1".$postkey.$postvalue;
			// loop
			if ('acf' == $postkey || 'acf-block' == substr( $postkey, 0, 9 ) ) {
			//	echo "2";
				foreach( $postvalue as $key => $value ) {
					//echo "3"."|".$postkey."|".$key."|".$value."|";
					// vars
					$field = acf_get_field( $key );
					//OLD $input = $input_prefix . '[' . $key . ']';
					///NEW to handle block prefixes
					$input = $postkey . '[' . $key . ']';
					
					
					// bail early if not found
					if( !$field ) continue;
					//echo "found";
					
					// validate
					acf_validate_value( $value, $field, $input );		
				}
			}
		}	
	} else {
		foreach( $values as $key => $value ) {

			// vars
			$field = acf_get_field( $key );
			$input = $input_prefix . '[' . $key . ']';
			
			// bail early if not found
			if( !$field ) continue;
			//echo "found";
			
			// validate
			acf_validate_value( $value, $field, $input );		
		}

	}
}

**Comment by todd-uams** [comment]
I can't seem to get Elliot's version to work. I've tried tweaking it a little, but I can't seem to get anything to stop the publish/update process. What am I missing?

**Comment by elvismdev** [comment]
I had the same experience as @todd-uams . Tried the snippet posted by @elliotcondon  a few days ago but no luck.

How should we add that to our current existing WP projects?

**Comment by elliotcondon** [comment]
Hi @todd-uams amd @elvismdev.

Thanks for the feedback. Would you be able to perform some basic JS debugging through the code to better articulate what is working / failing?

**Comment by todd-uams** [comment]
I'd be happy to. Is there something specific.

When I initially tried inserting the code, as is. I received a $ is not a function error. I tried changing the $('#editor') to JQuery('#editor'). This got rid of the error, but I'm not getting any response in Gutenberg. I've tried using it as an enqueued file and adding directly to the admin footer.

**Comment by elliotcondon** [comment]
@todd-uams Thanks for the feedback. I've just updated the example to take care of the `$` !== `jQuery` issue.

By "not getting any response", are you referring to the AJAX call? Or the modified `editor.savePost` function? If you could debug the JS, that would be super helpful to understand which parts are and aren't working for you.

**Comment by todd-uams** [comment]
I created a new clean install with just ACF Pro. For verification, I installed Gutenberg after testing to verify that it wasn't required. The results were the same. The block is just 4 fields. Text (required), True/False, Text Area (required), and a select field. The block just displays the information from the fields. 

So, the new code doesn't have any errors, but it still won't throw any validation errors. I have two fields that are required. If I run the field group outside of Gutenberg, it throws the standard validation errors. When I use it as a block, it just saves the data with empty fields. There are no console errors with the save, so it isn't an outside error.

I added a few console log lines to see if everything is firing as expected. The acf gutenbergValidation model is running. There is a single firing of `!valid` while it waits for acf Ajax (I believe). The success state fires next, followed by the complete state. I grabbed the data from the complete state, both for `$form` and `validator`. Both fire without errors.

@elliotcondon I'm not sure where else to look. I'd be happy to send any other information along.

**Comment by altendorfme** [comment]
Hi @elliotcondon, I added the code (https://github.com/AdvancedCustomFields/acf/issues/113#issuecomment-616949241) in 'acf-input.min.js', just for a test and it worked perfectly.

![Capture](https://user-images.githubusercontent.com/1530032/82264162-c4f89600-993a-11ea-9b55-e9a4d7de5ca2.gif)

**Comment by elliotcondon** [comment]
@altendorfme Thanks for the feedback!

**Comment by elliotcondon** [comment]
@todd-uams Please note this validation improvement is not for "ACF Blocks". The code mentioned above will validate metaboxes.

**Comment by todd-uams** [comment]
Sorry for the mixup. The code **did** work for metaboxes on a Gutenberg page. Thanks

**Comment by todd-uams** [comment]
@elliotcondon I'll look at @benfrem 's code for a potential solution for ACF Blocks.

**Comment by elliotcondon** [comment]
@todd-uams Fantastic! Block validation will be coming soon, but only after metabox validation is considered stable :)

**Comment by ggedde** [comment]
I can confirm that a mixture of the code above works great when using 'edit' mode. It also works when in 'preview' mode and editing a specific block, but only validates that block and not the others. Thanks!

However, I am still looking for a way to get this working in 'preview' mode for all blocks. @elliotcondon I know your working on a solution (Thank You), but since we don't have an ETA
is there any way that I can get all the fields and their values with the ACF JS API or with Gutengergs JS API so I can just perform a basic validation and loop through them and if the 'required' attribute is set and value is empty then I can set a notice. I am not really needing a way to open the block settings.  I think I will just use some css to add a red outline around the wp-block that has the error so the user knows which block to edit. So a way to reference the error and block would be helpful. The locking and preventing updates and publish is working.  I just need to validate all of the fields, but I find it difficult to figure out the gutenberg api and what options there are.

**Comment by ggedde** [comment]
@elliotcondon I have downloaded v5.9.0-beta1.
The Preview mode Validation does work, but only when viewing the block being edited and when the Settings window is open.  If I start to edit another block with passing validation and Publish the Page no Validation Errors get reported for the blocks that do not pass Validation. Also if I close the Settings window then no Validation errors get show.  It should still report the Validation Errors and prevent the post from being updated.

Also I think when there is an error on a block there should be a class added to the "wp-block" tag or the "acf-block-component" tag like 'acf-validation-error' and some sort of error indicator like a red outline etc. That way the User can see which blocks have errors and click on them to resolve.  Probably only needed for 'preview' mode or 'auto' mode when block is in preview state.  I am writing a custom class for my clients so I will most likely override the styles for my own and that is why a css class is important for me.

Let me know if there is a better place to report issues for v5.9.0-beta
Thanks

**Comment by elliotcondon** [comment]
Hi @ggedde. Please note the "Gutenberg Validation" feature in 5.9.0-beta1 is for metaboxes only - not "ACF Blocks". 

Block validation will be coming soon in a future update 👍.

**Comment by andjoen** [comment]
@elliotcondon Any updates on block validation?

**Comment by elliotcondon** [comment]
Hi @andjoen 👋. Block validation is planned or a "post 5.9" release. For now, we will focus on just the entire post/page validation. Looking forward to the challenge when we do dip our toes into that pond!

**Comment by ggedde** [comment]
@elliotcondon 

> Block validation is planned or a "post 5.9" release

I have seen a few updates to 5.9, but nothing mentions this issue. Do you have an idea as to when you might start working on this issue for 5.9.x.

I am willing to test anything if needed.
Thanks

**Comment by elliotcondon** [comment]
Thanks @ggedde. No updates just yet.

**Comment by StudioAl** [comment]
@mattgrshaw, yes, strange — I just went back to verify and, and now the post/page field validation is working. I'm not sure what the issue was for me at the time that prevented it from working — I was definitely on the latest version of the plugin.

**Comment by vignitres** [comment]
Still no solution for this. Has WordPress said anything about this? It's a very important feature...

**Comment by alexandre-tobia** [comment]
Hi @elliotcondon any updates about this ? 

We are talking about this bug since 2019 and still not working it's a major issue. Your page https://www.advancedcustomfields.com/resources/known-issues/ talking about it say : "Status: Similar to above, this feature is no longer working and is scheduled to be fixed in ACF 5.8." We are in 5.12.2... what's going on ? 

I know Gutenberg is very involved in this problem and I know you haven't found an "elegant" solution but can't you make a temporary "less elegant" fix in the mean time?

I don't want to sound like an angry guy but we pay for this plugin, I can understand that some things take time, but we are talking about more than 2 years on a major problem.

Thanks

**Comment by isuke01** [comment]
@alexandre-tobia I think the issue is "I know Gutenberg is very involved in this problem" that sentence is not true. The Gutenberg team is holding this back, they have to fix issue in their end ;/

**Comment by alexandre-tobia** [comment]
@isuke01 Thanks for your answer Sorry if my message was rude, not intented. Why not add thoses "temp" solution in the acf plugin ? it's better than nothing, no ? I don't see any update from gutenberg workign on this since 2020, so if acf don't include this temp solution i think we will never reach the end of the tunnel...

**Comment by ggedde** [comment]
Sorry for the rant,
If this is an issue mainly with Gutenberg, can we at least get some feedback on the conversations that are going on with Gutenberg's team and ACF's team. Can you provide a link to the Gutenberg Issue for this as I am not finding it? Or other links that we can use for reference and possibly support? What is the ETA for a solution?

@benbowler, have you found a solid solution yet?

it seems like this Issue is just getting ignored and it is a vital problem in my opinion. My client's keep asking, can you make that field required? Then I have to give them this long explanation. It honestly sounds like, to them, that I don't know what I am doing. "Why can't you just make the field required?" they say, and I don't blame them. I have already been looking at different solutions and have started migrating to building custom Gutenberg Blocks, but they are much more difficult to manage and create. The main issue with them is that they are much more difficult for clients to manage. It almost is like they need to go through a dedicated course to learn it. Not to mention keeping up to date on all the changes that happen on every WP update. It doesn't work for most clients.

ACF Block's is the way to go and I appreciate the work and dedication that Elliot has provided and that Delicious Brains has taken on. We just need this last feature to tie it all in.

I am willing to test / research / and work on this issue with others to find a solid solution. With the Temp solutions above it gets us 70% of the way there. I have worked on this, but I am not familiar with ACF's and Gutenberg's underlining js code and need help with getting each block's fields and submitted values and checking for required and conditionals, then reference the ACF field settings. I am able to do this with an open edited ACF block, but having a difficult time with the others. If you have some insights to that, then please let me know.

Thanks

**Comment by TheDesignerXD** [comment]
Any update? Why has Wordpress not added a simple hook before saving forms on the admin side that ACF can hook into? This seems like a simple simple simple fix. And over 4 years of waiting seems like a ridiculous amount of time to wait for such a small addition. During this time we have seen both ACF, and Wordpress go through countless updates, and new {insert year here} themes (that nobody I know uses) but no new simple little hook prior to saving forms on the admin side. For ACF being one of Wordpress' #1 plugins it is really surprising to me how little time or effort they put into helping ACF succeed when it was something WP did that borked the whole thing.

**Comment by christianMiguez** [comment]
Hy ther, any of you have solved or hacked this? I'm trying to require a field if a block is present, but validation doesn't work

**Comment by nikolowry** [comment]
Are there technical reasons for why @BenFremer's PR submitted at https://github.com/AdvancedCustomFields/acf/pull/751 hasn't been merged? 

We've followed @danielegiorgis example located at https://github.com/AdvancedCustomFields/acf/issues/752#issuecomment-1584205909 and are successfully validating using the `acf/validate_save_post` hook. 

It be great to get some clarification on why this hasn't officially been integrated -- trying to avoid any "known unknown" edge-case scenarios.

**Comment by 627426p** [comment]
Is there any update regarding this issue? What is the current status? Does it still lie with wordpress/core or gutenberg? What are the relevant tickets/issues so we can add our support there? 

Its absolutely incomprehensible why this critical issue doesnt have at least a functioning workaround after **more than 4 years**.

**Comment by landwire** [comment]
Just encountered that issue. Anyone has a solution for this?

**Comment by MattOndo** [comment]
Also just came across this issue, and it's a pretty big one. I'm surprised it's still lingering...

**Comment by eballeste** [comment]
about 5 years later, it's finally happening!

**Comment by isuke01** [comment]
Isn't it fixed yet?
Wasn't this introduced into ACF https://github.com/AdvancedCustomFields/acf/issues/113#issuecomment-616949241 ?

**Comment by ggedde** [comment]
I came found this post recently 
https://www.advancedcustomfields.com/blog/acf-6-3-0-released/#acf-blocks-validation
And thought WOW about time. I guess I just missed the update.

Well, I went to some of my projects and made sure they were on 6.3 and they already were. Most are all at 6.3.11
BUT, validation is still not working. 

Turns out it does work (ish) if you use the new block.json format to register your blocks, but it does not work if you register them with PHP via `acf_register_block_type` which is also deprecated.

I say work(ish) because it does have some bugs which I will need to add in a separate Issue, but good to know it is possible now.

**Comment by michaelw85** [comment]
Validation is still broken for me using 6.4 and `acf_register_block_type`, this seems to solve it though: https://github.com/AdvancedCustomFields/acf/issues/752#issuecomment-1584205909
